### PR TITLE
MG-42: model_details ETL

### PR DESCRIPTION
**Pull request by beatrizsaldana**
See Jira ticket [MG-42](https://sagebionetworks.jira.com/browse/MG-42) for details on expected behaviors.

[MG-42]: https://sagebionetworks.jira.com/browse/MG-42?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Problem
The Model AD project needed a new transformation module to process model details data from multiple source files into a structured JSON format. The data needed to be transformed according to specific requirements:
1. Join multiple source files (model_info, allele_info, pathology, biomarkers, human_transgene_allele_map) on the model column
2. Apply specific column renames and value transformations
- measure -> evidence_type
- ageDeath -> age
- measurement -> value
3. Handle special cases like human transgene allele mapping
4. Convert comma-delimited strings to lists for certain fields
5. Format biomarker values with HTML entities - Biomarker measure (pre-transform in source file) aka evidence_type (post-transform in output file) values use &beta; entity codes, instead of beta string literals
6. Standardize text formatting (e.g., Initial Caps for sex and tissue values)

## Solution
1. Created new transformation module `model_details.py`:
   - Implemented `transform_model_details()` as the main transformation function
   - Added helper functions for specific transformations:
     - `prepare_biomarker_pathology()` for standardizing biomarker/pathology data
     - `proccess_biomarker_pathology()` for grouping measurements
     - `process_genetic_info()` for handling human transgene allele mapping
   - Added comprehensive error handling with clear error messages
   - Implemented type conversion and data cleaning

2. Added `list_to_json()` to `load.py`:
   - New utility function to convert list of dictionaries to JSON
   - Supports proper handling of numpy types
   - Maintains consistent formatting with other JSON conversion functions

3. Created comprehensive test suite:
   - Added `test_model_details.py` with test cases for:
     - Successful transformations with good data
     - Handling of missing data
     - Empty datasets
     - Extra columns
     - No matching human transgene alleles
     - Missing models in different source files
     - Extra models in different source files
     - Error cases for missing datasets/columns
   - Created test assets in `tests/test_assets/model_details/`:
     - Input CSV files for various test scenarios
     - Expected output JSON files
     - Organized into input/output directories

Key design decisions:
1. Error Handling:
   - Used ValueError with descriptive messages instead of KeyError
   - Added input validation at the start of transformation
   - Preserved original errors as causes for debugging

2. Data Processing:
   - Created copies of input DataFrames to avoid modifying originals
   - Implemented case-insensitive merging for gene names
   - Standardized text formatting throughout the pipeline
   - Maintained consistent data types for IDs

3. Testing Strategy:
   - Comprehensive test coverage for all transformation scenarios
   - Separate test cases for each helper function
   - Validation of both successful and error cases

## Testing:
1. Automated Tests:
   - All test cases pass successfully
   - Coverage includes:
     - Basic transformations
     - Edge cases (empty data, missing data)
     - Error handling
     - Data type conversions
     - Text formatting
     - List conversions

2. Test Assets:
   - Created input CSV files for all test scenarios
   - Generated expected output JSON files
   - Organized test files in a clear directory structure
   - No sensitive data included in test files

### CODE REVIEW

File: src/agoradatatools/process.py [hunk 1]
```diff
@@ -166,6 +170,12 @@ def process_dataset(
             staging_path=staging_path,
             filename=dataset_name + "." + dataset_obj[dataset_name]["final_format"],
         )
+    if isinstance(df, list):
```
Comment (BWMac) [inline src/agoradatatools/process.py]: I think this should be `elif`

Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: OMG I totally missed this, thank you for taking a look!

File: src/agoradatatools/etl/load.py [hunk 2]
```diff
@@ -160,3 +160,21 @@ def dict_to_json(df: dict, staging_path: str, filename: str) -> str:
     json.dump(df_as_dict, temp_json, cls=NumpyEncoder, indent=2)
     temp_json.close()
     return temp_json.name
+
+
+def list_to_json(df: list, staging_path: str, filename: str) -> str:
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/load.py]: Nitpick: can we call the variable something other than `df` since it's a list? Though I guess the `dict_to_json` function does the same thing... hmm.

File: src/agoradatatools/etl/transform/model_details.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+def transform_model_details(
+    datasets: Dict[str, pd.DataFrame], dataset_name: str = "model_details"
+) -> List[Dict[str, Any]]:
+    """
+    Transforms the model_details datasets into a structured format.
+    See MG-42 for more details on expected structure: https://sagebionetworks.jira.com/browse/MG-42
+
+    Args:
+        datasets (Dict[str, pd.DataFrame]): Dictionary of dataset names mapped to their DataFrame.
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: Please add documentation about the `dataset_name` argument

File: src/agoradatatools/etl/transform/model_details.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    # Fill all NaN values with empty strings
+    biomarkers_df.fillna("", inplace=True)
+    human_transgene_allele_map_df.fillna("", inplace=True)
+    allele_info_df.fillna("", inplace=True)
+    model_info_df.fillna("", inplace=True)
+    pathology_df.fillna("", inplace=True)
+
+    # Transform text fields to Initial Caps
+    biomarkers_df["sex"] = biomarkers_df["sex"].str.capitalize()
+    biomarkers_df["tissue"] = biomarkers_df["tissue"].apply(
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: You should be able to use `x.title()` to convert to title case, instead of what's in the lambda function. Regarding the `isinstance(x, str)`, is there anything in the columns that is NOT a string? i.e. could this just be shortened to 
```
.apply(lambda x: x.title())
```
or
```
or .apply(title) <- not sure if that's valid
```
?

File: src/agoradatatools/etl/transform/model_details.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        lambda x: " ".join(word.capitalize() for word in x.split())
+        if isinstance(x, str)
+        else x
+    )
+
+    pathology_df["sex"] = pathology_df["sex"].str.capitalize()
+    pathology_df["tissue"] = pathology_df["tissue"].apply(
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: Same

File: src/agoradatatools/etl/transform/model_details.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                "modified_gene": allele_row["gene"],
+                "ensembl_id": allele_row["gene_ensembl_id"],
+                "allele": allele_row["allele"],
+                "allele_type": allele_row["allele_type"],
+                "mgi_allele_id": allele_row["mgi_allele_id"],
+            }
+
+            # If it's a human transgene, replace the ensembl_id with the human one
+            if (
+                allele_row["mgi_allele_id"]
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: There _might_ be a way to simplify this code a little bit by doing a full join (or left join?) of `allele_info_df` and `human_transgene_allele_map_df` on `mgi_allele_id` outside this for loop, pulling `allele_row` from that merged df, and checking if the gene symbol is NA or not. But without playing around I can't say for sure if that will work.

File: src/agoradatatools/etl/transform/model_details.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                "evidence_type": evidence_type,
+                "tissue": tissue,
+                "age": f"{age} months",
+                "units": group["units"].iloc[0],
+                "data": [],
+            }
+
+            # Add individual data points
+            for _, row in group.iterrows():
+                data_point = {
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: Another maybe -- If you use 
```
row[["genotype", "sex", "individual_id", "value"]].to_dict(orient="records")
```
instead of iterating with the for loop, this should give you a list of dicts with the proper keys and a single value for each key. If that works, you can also apply that to anywhere else applicable.

Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_details.py]: Addendum: I wonder if, by merging some of the data frames at the beginning of this function by `model`, you can kind of... mass do a bunch of this function with `.to_dict()` and eliminate some for loops inside here? Again, haven't played around, just thinking of ways to make the function a little shorter and easier to follow.

File: tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv [hunk 8]
```diff
@@ -0,0 +1 @@
+individual_id,specimen_id,model,type,measurement,units,age_death,tissue,sex,genotype
```
Comment (jaclynbeck-sage) [inline tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv:1]: I'm not sure we really need to test for this case

File: tests/test_assets/model_details/input/model_details_allele_info_good_test_input.csv [hunk 9]
```diff
@@ -0,0 +1,6 @@
+model,gene,mgi_gene_id,gene_ensembl_id,allele,allele_type,mgi_allele_id
```
Comment (jaclynbeck-sage) [inline tests/test_assets/model_details/input/model_details_allele_info_good_test_input.csv:1]: Can you add a test file for missing data for this data set? Unless there's already something missing in this file, hard to tell.

File: tests/test_assets/model_details/input/model_details_human_transgene_allele_map_good_test_input.csv [hunk 10]
```diff
@@ -0,0 +1,4 @@
+mgi_allele_id,gene_symbol,ensembl_id
```
Comment (jaclynbeck-sage) [inline tests/test_assets/model_details/input/model_details_human_transgene_allele_map_good_test_input.csv]: Similarly, add test for missing values

File: tests/test_assets/model_details/input/model_details_pathology_empty_input.csv [hunk 11]
```diff
@@ -0,0 +1 @@
+individual_id,specimen_id,model,type,measurement,units,age_death,tissue,sex,genotype
```
Comment (jaclynbeck-sage) [inline tests/test_assets/model_details/input/model_details_pathology_empty_input.csv:1]: Probably don't need to test for this

File: tests/test_assets/model_details/output/model_details_transform_no_human_match_output.json [hunk 12]
```diff
@@ -0,0 +1,286 @@
+[
```
Comment (jaclynbeck-sage) [inline tests/test_assets/model_details/output/model_details_transform_no_human_match_output.json:1]: Testing for no matches is a great idea. If you haven't done this already, can you make sure that that there is at least one test case where:
1. Each source file _doesn't_ have a row for at least one model in the main `model_info` file. Change which model is missing between the files so that there isn't just one row in the final list with no info. 
2. Each source file has an _extra row_ with a model that isn't in `model_info`

Just to make sure things are handled correctly in each case. I can easily see this happening in real life as new models get added.

File: tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv (hunk 8 above)
Comment (jaclynbeck-sage) [reply tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv:1]: Update: Oh I see how you're using this. You can probably leave this in, or you can omit the same model(s) from "good" biomarkers and pathology files to test the same thing.

File: tests/test_assets/model_details/input/model_details_pathology_empty_input.csv (hunk 11 above)
Comment (jaclynbeck-sage) [reply tests/test_assets/model_details/input/model_details_pathology_empty_input.csv:1]: Update: See my comment on the biomarkers empty file.

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
This is great! I didn't realize what a huge amount of work this needed to get it working.

I have some (optional) suggestions for shortening the transform, and also some additional things to test for.

File: src/agoradatatools/etl/transform/model_details.py (hunk 3 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_details.py]: Update: SonarCloud claims this argument is unused. I guess it can be removed.

File: src/agoradatatools/etl/load.py (hunk 2 above)
Comment (BWMac) [reply src/agoradatatools/etl/load.py]: Agreed. Just because we did something weird in the past doesn't mean we have to keep up the weirdness :sweat_smile:. The typing can also be more specific, i.e. `list[dict[str,str]]` or whatever the case may be.

File: src/agoradatatools/etl/load.py [hunk 13]
```diff
@@ -160,3 +160,21 @@ def dict_to_json(df: dict, staging_path: str, filename: str) -> str:
...
+        staging_path (str): Path to staging directory
+        filename (str): name of JSON file to be created
+
+    Returns:
+        str: Returns a string containing the name of the new JSON file
+    """
+
+    temp_json = open(os.path.join(staging_path, filename), "w+")
+    json.dump(df, temp_json, cls=NumpyEncoder, indent=2)
+    temp_json.close()
```
Comment (BWMac) [inline src/agoradatatools/etl/load.py]: NIT: It's recommended to do something like 
```
    with open(os.path.join(staging_path, filename), "w+") as temp_json:
        json.dump(df, temp_json, cls=NumpyEncoder, indent=2)
    return os.path.join(staging_path, filename)
```
because its safer given that the file is closed even if an exception were to occur

File: src/agoradatatools/etl/transform/model_details.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+import pandas as pd
+from typing import Any, Dict, List
+
+
+def transform_model_details(
+    datasets: Dict[str, pd.DataFrame], dataset_name: str = "model_details"
+) -> List[Dict[str, Any]]:
+    """
+    Transforms the model_details datasets into a structured format.
+    See MG-42 for more details on expected structure: https://sagebionetworks.jira.com/browse/MG-42
```
Comment (BWMac) [inline src/agoradatatools/etl/transform/model_details.py]: We probably don't want to link to Jira tickets within our codebase. I would just describe the expected functionality here so developers can see it more readily.

File: src/agoradatatools/etl/transform/model_details.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
+"""
+This module contains the transformation logic for the model_details datasets.
+This is for the Model AD project.
+"""
+import pandas as pd
+from typing import Any, Dict, List
+
+
+def transform_model_details(
```
Comment (BWMac) [inline src/agoradatatools/etl/transform/model_details.py]: This function is pretty long and has a lot of logic for handling different dataframes all included. Do you think it would be a good idea to break this up with helper functions? Anything that needs to be applied to all dataframes could be grouped together while anything specific could be separated out for each df type. Similar to some of Jaclyns' comments, I would also try to reduce the amount of repeated code here as much as possible too.

File: src/agoradatatools/etl/transform/model_details.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+            "jax_id": model_row["jax_id"],
+            "alzforum_id": model_row["alzforum_id"],
+            "genotype": model_row["genotype"],
+            "aliases": model_row["aliases"],
+            "genetic_info": genetic_info,
+            "biomarkers": model_biomarkers,
+            "pathology": model_pathology,
+        }
+
+        result.append(model_entry)
```
Comment (BWMac) [inline src/agoradatatools/etl/transform/model_details.py]: Since this is the only entity within the list, I would just do

```suggestion
        result = [
            {
            "model": model_name,
            "matched_controls": model_row["matched_controls"],
            "model_type": model_row["model_type"],
            "contributing_group": model_row["contributing_group"],
            "study_synid": model_row["study_synid"],
            "rrid": model_row["rrid"],
            "jax_id": model_row["jax_id"],
            "alzforum_id": model_row["alzforum_id"],
            "genotype": model_row["genotype"],
            "aliases": model_row["aliases"],
            "genetic_info": genetic_info,
            "biomarkers": model_biomarkers,
            "pathology": model_pathology,
            }
        ]
```

File: tests/transform/test_model_details.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    fail_test_data = [
+        (
+            # Fail with missing biomarkers dataset
+            {
+                "human_transgene_allele_map": "model_details_human_transgene_allele_map_good_test_input.csv",
+                "allele_info": "model_details_allele_info_good_test_input.csv",
+                "model_info": "model_details_model_info_good_test_input.csv",
+                "pathology": "model_details_pathology_good_test_input.csv",
+            },
+            KeyError,
```
Comment (BWMac) [inline tests/transform/test_model_details.py]: I think the fact that you are testing for `KeyError`s is a pretty good sign that there should be more error handling in the function itself. If another developer or a user were trying to run the pipeline and kept getting these errors it wouldn't be immediately clear what the issue was and how to fix it quickly.

**Review by BWMac — CHANGES_REQUESTED** [review]

**Comment by BWMac** [comment]
@beatrizsaldana One additional point. As a reviewer, having more details about your implementation in addition to just the Jira ticket would be helpful. 

On DPE we have adopted the Problem/Solution/Testing framework using this [PR template](https://github.com/Sage-Bionetworks-Workflows/nf-synapse/blob/main/.github/pull_request_template.md) (not saying we need to adopt the same norm in this repository, but just as an example). Having the context from a Jira ticket for what needed to be done is certainly helpful, but the additional context around what you actually ended up doing to meet the AC of a ticket including any tradeoffs or decisions that you made in the process and how you approached testing your changes would help too.

File: src/agoradatatools/etl/transform/model_details.py (hunk 16 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/model_details.py]: I think there is more than one model in the real dataset.

Comment (BWMac) [reply src/agoradatatools/etl/transform/model_details.py]: Oh I see I misread the indenting.

File: tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv (hunk 8 above)
Comment (beatrizsaldana) [reply tests/test_assets/model_details/input/model_details_biomarkers_empty_input.csv:1]: I'm going to leave it :)

File: tests/test_assets/model_details/input/model_details_pathology_empty_input.csv (hunk 11 above)
Comment (beatrizsaldana) [reply tests/test_assets/model_details/input/model_details_pathology_empty_input.csv:1]: I'm going to leave it :)

File: tests/test_assets/model_details/output/model_details_transform_no_human_match_output.json (hunk 12 above)
Comment (beatrizsaldana) [reply tests/test_assets/model_details/output/model_details_transform_no_human_match_output.json:1]: Addressed in recent commit

File: src/agoradatatools/etl/transform/model_details.py [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+            }
+            result.append(model_entry)
+
+        return result
+
+    except Exception as e:
+        raise ValueError(
+            f"Error processing model details: {str(e)}. "
+            "Please check that all input datasets are properly formatted and contain the required data."
+        ) from e
```
Comment (BWMac) [inline src/agoradatatools/etl/transform/model_details.py]: Doing this will prevent us from seeing the specific error thrown when processing and potentially mischaracterize errors that aren't ValueError. Do you think this is needed, or should we just let the actual error go through to the higher level error catcher we already have?

File: src/agoradatatools/process.py [hunk 19]
```diff
@@ -316,12 +324,7 @@ def process_all_files(
                 if dataset_report.gx_failures:
                     raise ADTDataValidationError(dataset_report.gx_failure_message)
         except Exception as e:
-            import traceback
-
-            error_message = (
-                f"{list(dataset.keys())[0]}:\n {str(e)}\n{traceback.format_exc()}"
-            )
-            error_list.append(error_message)
+            error_list.append(f"{list(dataset.keys())[0]}: " + str(e).replace("\n", ""))
```
Comment (BWMac) [inline src/agoradatatools/process.py]: Do you still need to merge dev into this branch?

File: src/agoradatatools/etl/transform/model_details.py (hunk 18 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/model_details.py]: Ahh true! I think we should let the actual error go through.

File: src/agoradatatools/process.py (hunk 19 above)
Comment (jaclynbeck-sage) [inline src/agoradatatools/process.py]: Does this interfere with Brad's changes to error printing at all?

Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Not sure why the merge didn't fix this. But I fixed it manually :)

Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Not sure why the merge didn't fix this. But I fixed it manually :)

File: tests/test_assets/model_details/input/model_details_human_transgene_allele_map_good_test_input.csv (hunk 10 above)
Comment (beatrizsaldana) [reply tests/test_assets/model_details/input/model_details_human_transgene_allele_map_good_test_input.csv]: Hmm I thought I did this but did not resolve the conversation. Need to look into it.

File: src/agoradatatools/etl/transform/model_details.py (hunk 18 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/model_details.py]: Removed the entire try except and kept the input validation. We should see specific errors if any inputs or their expected columns are missing. If a different error arises, we will see it as is.

**Review by BWMac — APPROVED** [review]
LGTM! 🔥

File: src/agoradatatools/etl/transform/model_details.py [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+    df["tissue"] = df["tissue"].str.title()
+
+    # Replace 'beta' with '&beta;' in biomarker types
+    df["type"] = df["type"].str.replace("beta", "&beta;")
+
+    # Rename columns
+    return df.rename(columns={"type": "evidence_type", "measurement": "value"})
+
+
+def proccess_biomarker_pathology(
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: Nitpick: should be "process" with one c

File: src/agoradatatools/etl/transform/model_details.py [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+    Args:
+        human_transgene_allele_map_df (pd.DataFrame): The DataFrame containing the human transgene allele information.
+        model_alleles (pd.DataFrame): The DataFrame containing the model allele information.
+
+    Returns:
+        List[Dict[str, Any]]: A list of dictionaries containing the processed gene information.
+    """
+    # Rename columns to match
+    human_df = human_transgene_allele_map_df.rename(
+        columns={"gene_symbol": "gene", "ensembl_id": "human_ensembl_id"}
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: Are these renames that can be done with "column_rename" in the config file?

File: src/agoradatatools/etl/transform/model_details.py [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+            "alzforum_id": model_row["alzforum_id"],
+            "genotype": model_row["genotype"],
+            "aliases": model_row["aliases"],
+            "genetic_info": genetic_info,
+            "biomarkers": model_biomarkers,
+            "pathology": model_pathology,
+        }
+        result.append(model_entry)
+
+    return result
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py:256]: Nice, this is so much cleaner and easier to read!

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
Almost there, I just have two nitpicks!

Great job cleaning up the transform :)

File: src/agoradatatools/etl/transform/model_details.py (hunk 20 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/model_details.py]: Thank you for catching this!

**Review by jaclynbeck-sage — APPROVED** [review]
Excellent, looks good!
